pkg/networkservice/chains/nsmgr/suite_test.go

```diff
@@ -528,3 +725,201 @@ func (s *nsmgrSuite) Test_ShouldCleanAllClientAndEndpointGoroutines() {
+
+		matches = append(matches, match)
+	}
+
+	if count > 1 {
+		// match with empty source selector must be the last
+		match := &registry.Match{
+			Routes: []*registry.Destination{
+				{
+					DestinationSelector: map[string]string{
+						step: fmt.Sprintf("%v", 0),
+					},
+				},
+			},
+		}
+		matches = append(matches, match)
+	}
+
+	return &registry.NetworkService{
+		Name:    "test-network-service-linear",
+		Matches: matches,
+	}
+}
+
+func multiLabelNS() *registry.NetworkService {
+	return &registry.NetworkService{
+		Name: "test-network-service",
+		Matches: []*registry.Match{
+			{
+				SourceSelector: map[string]string{
+					labelA: "a",
+					labelB: "b",
+				},
+				Routes: []*registry.Destination{
+					{
+						DestinationSelector: map[string]string{
+							labelA: "aa",
+							labelB: "bb",
+						},
+					},
+				},
+			},
+			{
+				SourceSelector: map[string]string{
+					labelA: "aa",
+					labelB: "bb",
+				},
+				Routes: []*registry.Destination{
+					{
+						DestinationSelector: map[string]string{
+							labelA: "aaa",
+							labelB: "bbb",
+						},
+					},
+				},
+			},
+			{
+				Routes: []*registry.Destination{
+					{
+						DestinationSelector: map[string]string{
+							labelA: "a",
+							labelB: "b",
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func treeNS() *registry.NetworkService {
+	return &registry.NetworkService{
+		Name: "test-network-service",
+		Matches: []*registry.Match{
+			{
+				SourceSelector: map[string]string{
+					step: "aa",
+				},
+				Routes: []*registry.Destination{
+					{
+						DestinationSelector: map[string]string{
+							step: "ab",
+						},
+					},
+				},
+			},
+			{
+				SourceSelector: map[string]string{
+					step: "ab",
+				},
+				Routes: []*registry.Destination{
+					{
+						DestinationSelector: map[string]string{
+							step: "c",
+						},
+					},
+				},
+			},
+			{
+				SourceSelector: map[string]string{
+					step: "c",
+				},
+				Routes: []*registry.Destination{
+					{
+						DestinationSelector: map[string]string{
+							step: "d",
+						},
+					},
+				},
+			},
+			{
+				SourceSelector: map[string]string{
+					step: "ba",
+				},
+				Routes: []*registry.Destination{
+					{
+						DestinationSelector: map[string]string{
+							step: "bb",
+						},
+					},
+				},
+			},
+			{
+				SourceSelector: map[string]string{
+					step: "bb",
+				},
+				Routes: []*registry.Destination{
+					{
+						DestinationSelector: map[string]string{
+							step: "c",
+						},
+					},
+				},
+			},
+			{
+				Routes: []*registry.Destination{
+					{
+						DestinationSelector: map[string]string{
+							step: "aa",
+						},
+					},
+					{
+						DestinationSelector: map[string]string{
+							step: "ba",
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func additionalFunctionalityChain(ctx context.Context, clientURL *url.URL, clientName string, labels map[string]string, empty bool) []networkservice.NetworkServiceServer {
+	if empty {
```

Comment:
Probably it would be better to do this check outside of the method and just not call it in such case.

Reply:
done